[DI] Draft frontend plugins testing documentation (#21782)
backstage/backstage: docs/frontend-system/building-plugins/testing.md

---
id: testing
title: Frontend System Testing Plugins
sidebar_label: Testing
# prettier-ignore
description: Testing plugins in the frontend system
---

> **NOTE: The new frontend system is in a highly experimental phase**

# Testing Frontend Plugins

> NOTE: The new frontend system is in alpha, and some plugins do not yet fully implement it.

Utilities for testing frontend features and components are available in `@backstage/frontend-test-utils`.

## Testing React components

A component can be used for more than one extension, and it should be tested independently of an extension environment.

Use the `renderInTestApp` helper to render a given component inside a Backstage test app:

```tsx
import React from 'react';
import { screen } from '@testing-library/react';
import { renderInTestApp } from '@backstage/frontend-test-utils';
import { EntityDetails } from './plugin';

describe('Entity details component', () => {
  it('should render the entity name and owner', async () => {
    await renderInTestApp(<EntityDetails owner="tools" name="test" />);

    await expect(

>       screen.getByText('The entity "test" is owned by "tools"'),

Reviewer: ```suggestion
      screen.findByText('The entity "test" is owned by "tools"'),
```

Reviewer 2: Shipping this :p

---
id: testing
title: Frontend System Testing Plugins
sidebar_label: Testing
# prettier-ignore
description: Testing plugins in the frontend system
---

> **NOTE: The new frontend system is in a highly experimental phase**

# Testing Frontend Plugins

> NOTE: The new frontend system is in alpha, and some plugins do not yet fully implement it.

Utilities for testing frontend features and components are available in `@backstage/frontend-test-utils`.

> ## Testing components

Reviewer 2: ```suggestion
## Testing React components
```

Imo, just to make sure that's clear 😁

A component can be used for more than one extension, and it should be tested independently of an extension environment.

Use the `renderInTestApp` helper to render a given component inside a Backstage test app:

```tsx
import React from 'react';
import { screen } from '@testing-library/react';
import { renderInTestApp } from '@backstage/frontend-test-utils';
import { EntityDetails } from './plugin';

describe('Entity details component', () => {
  it('should render the entity name and owner', async () => {
    renderInTestApp(<EntityDetails owner="tools" name="test" />);

    await expect(
      screen.getByText('The entity "test" is owned by "tools"'),
    ).resolves.toBeInTheDocument();
  });
});
```

> It's important to highlight that mocking APIs for components is different from mocking them for extensions. In the snippet below, we wrapped the component within a `TestApiProvider` for mocking the Catalog API:

Reviewer 2: ```suggestion
To mock [Utility APIs](../utility-apis/01-index.md) that are used by your component you can use the `TestApiProvider` to override individual API implementations. In the snippet below, we wrap the component within a `TestApiProvider` in order to mock the catalog client API:
```

Thinking we can keep this straight to the point of how to mock APIs for components and worry about extensions later.

```tsx
import React from 'react';
import { screen } from '@testing-library/react';
import {
  renderInTestApp,
  TestApiProvider,
} from '@backstage/frontend-test-utils';
import { stringifyEntityRef } from '@backstage/catalog-model';
import { CatalogApi, catalogApiRef } from '@backstage/plugin-catalog-react';
import { EntityDetails } from './plugin';

describe('Entity details component', () => {
  it('should render the entity name and owner', async () => {
    const catalogApiMock: Partial<CatalogApi> = {
      getEntityFacets: () =>
        Promise.resolve({
          facets: {
            'relations.ownedBy': [{ count: 1, value: 'group:default/tools' }],
          },
        }),

>     };

Author: TIL, thank you 🙌🏻

    const entityRef = stringifyEntityRef({
      kind: 'Component',
      namespace: 'default',
      name: 'test',
    });

>     renderInTestApp(

Author: Done 🙌🏻

      <TestApiProvider apis={[[catalogApiRef, catalogApiMock]]}>
        <EntityDetails entitRef={entityRef} />
      </TestApiProvider>,
    );

    await expect(
      screen.getByText('The entity "test" is owned by "tools"'),
    ).resolves.toBeInTheDocument();
  });
});
```

> ## Testing features

Reviewer 2: ```suggestion
## Testing extensions
```

Since a frontend feature is now `plugin | overrides` I'm thinking it might be a bit confusing to use it in this context too. Same thing below

Author: Makes sense, done 🙌🏻

To facilitate testing of frontend features, the `@backstage/frontend-test-utils` package provides a tester class which starts up an entire frontend harness, complete with a number of default features. You can then provide overrides for extensions whose behavior you need to adjust for the test run.

A number of features (frontend extensions and overrides) are also accepted by the tester. Here are some examples of how these facilities can be useful:

### Single extension

In order to test an extension in isolation, you simply need to pass it into the tester factory, then call the render method on the returned instance:

```tsx
import { screen } from '@testing-library/react';
import { createExtensionTester } from '@backstage/frontend-test-utils';
import { indexPageExtension } from './plugin';

describe('Index page', () => {
  it('should render a the index page', () => {

>     const tester = createExtensionTester(indexPageExtension);
> 
>     tester.render();

Reviewer 2: ```suggestion
    createExtensionTester(indexPageExtension).render();
```

Pending other possible versions of this API I think we should encourage the direct chain version of calling it for now. Same for the examples below

Author: Done 🙌🏻

    expect(screen.getByText('Index Page')).toBeInTheDocument();
  });
});
```

### Extension preset

There are some extensions that rely on other extensions existence, such as a page that links to another page. In that case, you can add more than one extension to the preset of features you want to render in the test, as shown below:

```tsx
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createExtensionTester } from '@backstage/frontend-test-utils';
import { indexPageExtension, detailsPageExtension } from './plugin';

describe('Index page', async () => {
  it('should link to the details page', () => {
    const tester = createExtensionTester(indexPageExtension);

    // Adding more extensions to the preset being tested
    tester.add(detailsPageExtension);
    tester.render();

>     expect(screen.getByText('Index Page')).toBeInTheDocument();

Reviewer: ```suggestion
    await expect(screen.findByText('Index Page')).toBeInTheDocument();
```

it's just a neat habit to wait for results in case the minute details of the app makes them appear a single tick later or so

Author: Done 🙌🏻

>     userEvent.click(screen.getByRole('link', { name: 'See details' }));

Reviewer: ```suggestion
    await userEvent.click(screen.getByRole('link', { name: 'See details' }));
```

userEvent things are always async!

>     await expect(screen.getByText('Details Page')).resolves.toBeInTheDocument();

Reviewer: ```suggestion
    await expect(screen.findByText('Details Page')).resolves.toBeInTheDocument();
```

Author: Done 🙌🏻

  });
});
```

### Mocking apis

If your extensions requires implementation of APIs that aren't wired up by default, you'll have to add overrides to the preset of features being tested:

```tsx
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createApiFactory } from '@backestage/core-plugin-api';
import {
  createExtensionOverrides,
  configApiRef,
  analyticsApiRef,
} from '@backstage/frontend-plugin-api';
import {
  createExtensionTester,
  MockConfigApi,
  MockAnalyticsApi,
} from '@backstage/frontend-test-utils';
import { indexPageExtension } from './plugin';

describe('Index page', () => {

>   it('should capture click events in anylitics', async () => {

Reviewer: ```suggestion
  it('should capture click events in analytics', async () => {
```

    // Mocking the analytics api implementation
    const analyticsApiMock = new MockAnalyticsApi();

    const analyticsApiOverride = createApiExtension({
      factory: createApiFactory({
        api: analyticsApiRef,
        factory: () => analyticsApiMock,
      }),
    });

    const tester = createExtensionTester(indexPageExtension);

    // Overriding the analytics api extension
    tester.add(analyticsApiOverride);

    tester.render();

>     userEvent.click(await screen.findByRole('link', { name: 'See details' }));

Reviewer: ```suggestion
    await userEvent.click(await screen.findByRole('link', { name: 'See details' }));
```

});
```

### Setting configuration

In the case that your extension can be configured, you can test this capability by passing configuration values as follows:

```tsx
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createExtensionTester } from '@backstage/frontend-test-utils';
import { indexPageExtension, detailsPageExtension } from './plugin';

describe('Index page', () => {
  it('should accepts a custom title via config', async () => {
    const tester = createExtensionTester(indexPageExtension, {
      // Configuration specific of index page
      config: { title: 'Custom index' },
    });

    tester.add(detailsExtensionPage, {
      // Configuration specific of details page
      config: { title: 'Custom details' },
    });

    tester.render({
      // Configuration specific of the instance
      config: {
        app: {
          title: 'Custom app',
        },
      },
    });

    await expect(
      screen.findByRole('heading', { name: 'Custom app' }),
    ).resolves.toBeInTheDocument();

    await expect(
      screen.findByRole('heading', { name: 'Custom index' }),
    ).resolves.toBeInTheDocument();

>     userEvent.click(screen.getByRole('link', { name: 'See details' }));

Reviewer: ```suggestion
    await userEvent.click(screen.getByRole('link', { name: 'See details' }));
```
